R/aes-calculated.r

```diff
@@ -36,7 +73,12 @@ is_dotted_var <- function(x) {
 is_calculated_aes <- function(aesthetics) {
   vapply(aesthetics, is_calculated, logical(1), USE.NAMES = FALSE)
 }
-
+is_mapped_aes <- function(aesthetics) {
```

Comment:
These semantics feel a bit sloppy to me because they allow users to affect the interpretation of an expression by using `mapped()` or `stage()` deep in the call, even though that applies to the whole expression, not subparts. It seems it would be cleaner to require the qualifier to be at top-level.

Then you could check for a `mapped()` call at top level, and bind `mapped` to a function that fails with an error message in the function-level of the data mask.

Reply:
On a side note, I find it confusing to have `is_` predicate return length != 1 vectors. I think an `is_` function should always be safe to use in `if ()` (length 1, no missing value). In rlang I use the `are_` prefix for vectorised predicates (but this convention wasn't picked up anywhere else I think).

Reply:
The whole `is_mapped` and `is_mapped_aes` setup is borrowed completely from the way `stat()` calls are handled. @Hadley do you think there is reason to rewrite them both?

Reply:
ok, so I'll provide simplified parsers for `mapped()` (or whatever it ends up being called) and `stage()`

Reply:
Do we agree that something like 
```r
is_scaled <- function(x) {
  is.call(x) && identical(x[[1]], quote(after_scale))
}
```
should be sufficient?

Reply:
Given that we use rlang, you can use `is_call(x, "after_scale")`

Reply:
oh, that is handy

Reply:
but doesn't work on quosures... @lionel- is that a bug? The documentation says it will match name on the res if `x` is a formula - shouldn't the same be applied to quosures?

Reply:
```r
t <- quo(test())
is_call(t)
#> [1] TRUE
is_call(t, "test")
#> [1] FALSE
```
```

Reply:
hmm... it doesn't work for formulas either
```r
t <- ~test()
is_call(t, "test")
#> [1] FALSE
```

Reply:
This is fixed in dev rlang, `is_call()` consistently returns `FALSE` on quosures, forcing you to special-case them as appropriate.

Reply:
will the new slang be out soon? 🙂

Reply:
Next week.

Reply:
great

Reply:
Just tried with GitHub rlang - still doesn't work... Can you confirm the above example still fails?

Reply:
I get the expected results:

```r
t <- quo(test())

is_call(t)
#> [1] FALSE

is_call(t, "test")
#> [1] FALSE
```

Reply:
Ah... I misunderstood... I thought that `FALSE` was an error and that it would be fixed with next release...

Reply:
so... would this be the correct way to test for whether a quosure is of a specific call?
```r
is_call(quo_squash(t), 'test')
```
or are there any gotchas to that approach?

Reply:
You should almost never use `quo_squash()`, its name is meant to be scary ;). `quo_get_expr()` is the way to extract an expression.

- If you know you have a quosure, use `quo_is_call()` or `is_call(quo_get_expr())`.

- If you'd like to automatically unwrap quosures and formulas (this can be tricky and cause unexpected results), use `is_call(get_expr())`.

Reply:
```r
get_expr(quote(foo()))
#> foo()

get_expr(~ foo())
#> foo()

get_expr(quo(foo()))
#> foo()
```